crates/store/re_chunk/src/iter.rs

```diff
@@ -1,29 +1,179 @@
+    /// and component.
+    pub fn iter_component_offsets(
+        &self,
+        component_name: &ComponentName,
+    ) -> impl Iterator<Item = (usize, usize)> + '_ {
+        let Some(list_array) = self.components.get(component_name) else {
+            return Either::Left(std::iter::empty());
+        };
+
+        let offsets = list_array.offsets().iter().map(|idx| *idx as usize);
+        let lengths = list_array.offsets().lengths();
+
+        if let Some(validity) = list_array.validity() {
+            Either::Right(Either::Left(
+                izip!(offsets, lengths)
+                    .enumerate()
+                    .filter_map(|(i, o)| validity.get_bit(i).then_some(o)),
+            ))
+        } else {
+            Either::Right(Either::Right(izip!(offsets, lengths)))
+        }
+    }
+
+    /// Returns an iterator over the raw arrays of a [`Chunk`], for a given component.
+    ///
+    /// See also:
+    /// * [`Self::iter_component`].
+    /// * [`Self::iter_primitive`].
+    pub fn iter_component_arrays(
+        &self,
+        component_name: &ComponentName,
+    ) -> impl Iterator<Item = Box<dyn ArrowArray>> + '_ {
+        let Some(list_array) = self.components.get(component_name) else {
+            return Either::Left(std::iter::empty());
+        };
+
+        Either::Right(list_array.iter().flatten())
+    }
+
+    /// Returns an iterator over the raw primitive values of a [`Chunk`], for a given component.
+    ///
+    /// This is a very fast path: the entire column will be downcasted at once, and then every
+    /// component batch will be a slice reference into that global slice.
+    /// Use this when working with simple arrow datatypes and performance matters (e.g. scalars,
+    /// points, etc).
+    ///
+    /// * [`Self::iter_component_arrays`].
+    /// * [`Self::iter_component`].
+    #[inline]
+    pub fn iter_primitive<T: arrow2::types::NativeType>(
+        &self,
+        component_name: &ComponentName,
+    ) -> impl Iterator<Item = &[T]> + '_ {
+        let Some(list_array) = self.components.get(component_name) else {
+            return Either::Left(std::iter::empty());
+        };
+
+        let Some(values) = list_array
+            .values()
+            .as_any()
+            .downcast_ref::<PrimitiveArray<T>>()
+        else {
+            if cfg!(debug_assertions) {
+                panic!("downcast failed for {component_name}, data discarded");
+            } else {
+                re_log::error_once!("downcast failed for {component_name}, data discarded");
+            }
+            return Either::Left(std::iter::empty());
+        };
+        let values = values.values().as_slice();
+
+        // NOTE: No need for validity checks here, `iter_offsets` already takes care of that.
+        Either::Right(
+            self.iter_component_offsets(component_name)
+                .map(move |(idx, len)| &values[idx..idx + len]),
+        )
+    }
+}
+
+// TODO(cmc): get rid of this
```

Comment:
This happens later in that PR train -- at the moment this is still used all over the place.